packages/pigeon/lib/dart_generator.dart

```diff
@@ -615,11 +667,23 @@ void generateTestDart(
   indent.writeln('import \'package:flutter/services.dart\';');
   indent.writeln('import \'package:flutter_test/flutter_test.dart\';');
   indent.writeln('');
-  // TODO(gaaclarke): Switch from relative paths to URIs. This fails in new versions of Dart,
-  // https://github.com/flutter/flutter/issues/97744.
-  indent.writeln(
-    'import \'${_escapeForDartSingleQuotedString(mainDartFile)}\';',
+  final String relativeDartPath =
+      path.Context(style: path.Style.posix).relative(
+    _posixify(dartOutPath),
+    from: _posixify(path.dirname(testOutPath)),
   );
+  late final String? packageName = _deducePackageName(dartOutPath);
+  if (!relativeDartPath.contains('/lib/') || packageName == null) {
+    // If we can't figure out the package name or the relative path doesn't
+    // include a 'lib' directory, try relative path import which only works in
+    // certain (older) versions of Dart.
+    indent.writeln(
+        'import \'${_escapeForDartSingleQuotedString(relativeDartPath)}\';');
+  } else {
+    final String path = relativeDartPath.replaceFirstMapped(
+        RegExp(r'.*/lib/(.*?)'), (Match match) => match.group(1).toString());
```

Comment:
Won't `(.*?)` always be empty?

Reply:
You know what, you might be right in theory.  In practice it's doing the (relatively) correct operation.  It's asserted to work in the unit test.  It's almost like there is an implied `$` at the end of the regex for some reason.  I added explicit `^` and `$` characters to avoid confusion.  I'll file a dart bug.

Reply:
I don't think there's a Dart bug here, what you wrote is just a weird way of writing an empty replacement. Replacing `.*/lib/(.*)$` with `'$1'` and replacing `.*/lib/` with `''` are the same thing.

Reply:
Oh yea, weird, I see now.  Funny how that worked out.